### PR TITLE
GPFS alltoallv

### PR DESCRIPTION
**Pull request by markalle**
porting an old fix we made in OMPI's copy of romio back to mpich

There was a problem with the Alltoallv in ADIOI_GPFS_Calc_others_req().
It was using sendbuf and sdisps[] to reference the various
my_req[x].offsets, but those are all malloced separately and could
thus be further apart then the integers in sdisps[] would allow.
So this PR copies all that data into a new send buf before the
Alltoallv, then copies it back out of a new recv buf.

I don't have a testcase that demonstrates the failure, but I did
at least run a GPFS test that called MPI_File_write_all() and
uses the Alltoallv().

Separately I also had to fix the GPFS romio build a little bit for it to
build and run at all.  That part is in the first commit

### CODE REVIEW

**Comment by markalle** [comment]
bot:retest

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/warnings

**Comment by gpaulsen** [comment]
It looks like this is one of three ROMIO PRs needed to fix an HDF5 issue on Lustre.  Please see https://github.com/open-mpi/ompi/issues/6871 for details.

**Comment by markalle** [comment]
I just added another commit (build environment related) to this PR.  In OMPI another romio is being pulled in and it didn't all build in that environment till I added a _GNU_SOURCE and a typecast to avoid a compiler error with -Wpedantic.

The _GNU_SOURCE is to make sure that aligned_alloc() is declared.  Or more specifically to make sure that the determination about aligned_alloc()'s status made at "configure" time matches what happens later when gpfs files are being built.  The way the failure happened was configure used _GNU_SOURCE and determined that aligned_alloc() was already declared, then in the gpfs files stdlib.h was included before _GNU_SOURCE was defined so aligned_alloc() wasn't defined.

**Comment by markalle** [comment]
I don't think I have permission to add a reviewer.  @roblatham00 / @hzhou can I get somebody to review?

The code change itself is for GPFS where an Alltoallv() was using oversized disps[] so I copied the same data into an array where the disps[] would be legal.  Then the other two commits are just small things that were keeping the GPFS build from building and/or fixes for different build environments

**Comment by hzhou** [comment]
I'll take a look

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c [hunk 1]
```diff
@@ -709,54 +699,57 @@ void ADIOI_GPFS_Calc_others_req(ADIO_File fd, int count_my_req_procs,
...
+        sdispls[i] = scount_total;
+        scount_total += scounts[i];
+        rdispls[i] = rcount_total;
+        rcount_total += rcounts[i];
+    }
+    sbuf_copy_of_req_info = (ADIO_Offset *) ADIOI_Malloc(scount_total * sizeof(ADIO_Offset));
+    rbuf_copy_of_req_info = (ADIO_Offset *) ADIOI_Malloc(rcount_total * sizeof(ADIO_Offset));
+    for (i = 0; i < nprocs; i++) {
+        memcpy(sbuf_copy_of_req_info + sdispls[i] * sizeof(ADIO_Offset),
+               my_req[i].offsets, scounts[i] * sizeof(ADIO_Offset));
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c:737]: Should we concern about the extra memcpy overhead plus the extra memory footprint, considering some applications may do huge buffer operations.

There is MPI-4 API coming that adds `MPI_Alltoallv_c` to allow large count displs array. Maybe we can wait for that.

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs.h [hunk 2]
```diff
@@ -11,6 +11,18 @@
...
+ * This might be needed to get an aligned_alloc() declaration,
+ * at least it was on PPC64 RHEL 7.5.  The romio configure has
+ * a check that tries to identify NEEDS_ALIGNED_ALLOC_DECL, but
+ * at that point the conftest.c it has _GNU_SOURCE defined and
+ * that's why it decides it doesn't need to declare that function.
+ * But for _GNU_SOURCE to be effective it has to not only be
+ * defined but specifically be defined in front of includes like
+ * stdlib.h, so that makes it a little finicky
+ */
+#define _GNU_SOURCE
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_gpfs/ad_gpfs.h:24]: Feels like the proper solution is not to add `_GNU_SOURCE` in configure.

**Comment by markalle** [comment]
Thanks for the review.  About _GNU_SOURCE, here's a reproducer (fails RH7.5 anyway, passes on RH8):
```
% ./configure \
    --prefix=/smpi_dev/markalle/Projects/mpich/install \
    --with-device=ch4:ucx --with-file-system=nfs+ufs+gpfs \
    CFLAGS=-Werror=implicit-function-declaration
% cd src/mpi/romio
% make
```

Automake is definitely not my strong suit, but I've mostly traced what's happening.  The sequence of events that leads to _GNU_SOURCE being there when configure tests if aligned_alloc() needs declared is
1. src/mpi/romio/mpl/include/config.h.in has it
2. src/mpi/romio/mpl/configure.ac uses
   AC_CONFIG_HEADER[include/config.h]
   which puts those contents in confdefs.h
3. the declaration test is called in src/mpi/romio/mpl/configure.ac at
   PAC_FUNC_NEEDS_DECL([#include <stdlib.h>],aligned_alloc)
4. the PAC_FUNC_NEEDS_DECL test is defined in
   src/mpi/romio/mpl/confdb/aclocal_cc.m4
   and uses AC_TRY_COMPILE([$1 ...], ...)
   which uses the confdefs.h set up in AC_CONFIG_HEADER.

So I think fundamentally if src/mpi/romio/mpl/include/config.h is to be trusted, then that means anything using mpl is expected to be _GNU_SOURCE.  And it sort of is:  the config.h settings make their way into ./src/mpl/include/mplconfig.h and when you include mpl.h it includes mpl_base.h which includes mplconfig.h, so that's where it gets _GNU_SOURCE.  So at least from mpl.h's perspective it's including stdlib.h after it defines _GNU_SOURCE.

But an example failure is this include sequence where the things that happen before mpl.h are the problem (eg, they include stdlib.h before _GNU_SOURCE gets defined in mpl.h).

```
adio/ad_gpfs/ad_gpfs_close.c
  adio/ad_gpfs/ad_gpfs.h
    ./adio/include/adio.h
      ./adio/include/adioi.h
        ./mpl/include/mpl.h
          ./mpl/include/mpl_trmem.h
            (implicit declaration of function 'aligned_alloc')
```

I don't think I have a better solution than what I put in this PR.  To put mpl.h first would mean not just that adioi.h needs mpl.h at its top, but also adio.h needs aioio.h at its top etc because it has to magically know that adioi.h is going to be a consumer of mpl.h.

**Comment by markalle** [comment]
Okay I've switched to an alternate fix that's more of a sidestep of the problem than a fix exactly.

The reason the include order became a factor in potentially any .c/.h file is because the header was making MPL_aligned_alloc() a static function.  I'm hoping going non-static there is a small enough performance loss that the gain in control over the include order (only have to control it for one file now) is worth the change

**Comment by markalle** [comment]
And now I redid the Alltoallv too so it's capable of using MPI_Alltoall_c when it comes available.  Now instead of modifying the part that calls Alltoallv I have it going into a MY_Alltoallv() that uses vanilla MPI_Alltoallv() if the disps[] arrays are lucky enough to all fit in ints.  Then it does all the buffer copying from before if the disps[] aren't okay.

**Comment by roblatham00** [comment]
We definitely saw memory pressure on Blue Gene for some collective calls and this MPI_ALLTOALLV -- a memory allocation would fail and code would abort when trying to do certain collective operations.  "collective read of the same data" would do it but so too would some mesh workloads where the data was irregularly distributed among the participating processes . I have not seen that memory pressure on any other system -- then again, this optimization was only used on Blue Gene. 

It sounds like this fixes a problem in the wild, though, and it's easy enough to disable collectives or pretend the file is generic unix if someone runs into a memory allocation problem.

**Comment by markalle** [comment]
My notes aren't great on exactly what app I was using to hit the bug but I think it was a medium-sized setup, maybe 8 hosts with 40 ranks each or thereabouts.  It came from some customer so it was being hit in the wild.  I don't recall how big the alltoallv buffers were in the failure either.  I didn't think they were huge, I would have probably remembered that.

I guess I could make this fix more capable of dealing with high memory usage: it could allocate something moderate then step through the alltoallv data over multiple calls.  Is that worth pursuing, or are you saying that in cases where this fix results in too much memory use, the expectation is the user would take a different path?

**Comment by gpaulsen** [comment]
@hzhou - Hello.  Are the `mpich-warnings` something that needs to be resolved before this gets merged?  Are you waiting on anything from @markalle (on my team here at IBM)?  I ask because we're waiting to merge this same commit into a release branch in Open MPI: https://github.com/open-mpi/ompi/pull/8371.  Thanks.

**Comment by hzhou** [comment]
> @hzhou - Hello. Are the `mpich-warnings` something that needs to be resolved before this gets merged? Are you waiting on anything from @markalle (on my team here at IBM)? I ask because we're waiting to merge this same commit into a release branch in Open MPI: [open-mpi/ompi#8371](https://github.com/open-mpi/ompi/pull/8371). Thanks.

Sorry about the delay. No, it is not because of the warnings test. It is just that we need a few asks like yours to move it up on our radar :)  I'll review it shortly

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c [hunk 3]
```diff
@@ -360,7 +360,7 @@ void ADIOI_GPFS_WriteStridedColl(ADIO_File fd, const void *buf, int count,
             ADIOI_OneSidedWriteAggregation(fd, offset_list, len_list, contig_access_count, buf,
                                            datatype, error_code, firstFileOffset, lastFileOffset,
                                            currentValidDataIndex, fd_start, fd_end, &holeFound,
-                                           noStripeParms);
+                                           &noStripeParms);
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c:363]: There are two things going on in this commit (`making romio GPFS able to build and run again`). Can you split them?

File: src/mpl/src/mem/mpl_trmem.c [hunk 4]
```diff
@@ -1021,3 +1021,51 @@ void MPL_direct_free(void *ptr)
...
+
+    ret = posix_memalign(&ptr, alignment, size);
+    if (ret != 0)
+        return NULL;
+    return ptr;
+#else
+#error "MPL_DEFINE_ALIGNED_ALLOC defined but no underlying support function found - should not reach here."
+#endif
+}
+#endif /* #ifdef MPL_DEFINE_ALIGNED_ALLOC */
```
Comment (hzhou) [inline src/mpl/src/mem/mpl_trmem.c:1071]: This commit (MPL `aligned_alloc`) need be split into a separate PR. It is does not only impact ROMIO and it is orthogonal to other changes anyway.

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c [hunk 5]
```diff
@@ -769,3 +793,82 @@ void ADIOI_GPFS_Calc_others_req(ADIO_File fd, int count_my_req_procs,
...
+        }
+        for (i = 0; i < nranks; ++i) {
+            rdisps_int[i] = rdisps[i];
+        }
+        rv = MPI_Alltoallv(sbuf, scounts, sdisps_int, stype,
+                           rbuf, rcounts, rdisps_int, rtype, comm);
+        ADIOI_Free(sdisps_int);
+        return rv;
+    }
+
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c:846]: How about wrap the following code into the `else` branch?

**Review by hzhou — CHANGES_REQUESTED** [review]

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c (hunk 3 above)
Comment (hzhou) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c:363]: Ideally this commit should also move to separate PR as it is not related to the Alltoallv fix

Comment (markalle) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c:363]: Is that really better?  It turns the main PR into something that can't pass CI until a separate PR is checked in.  This way you at least have a PR that can be built and run without intra-PR dependencies.  I can change it, just you won't be able to test this PR or have it pass CI in the meantime

The fixes related to different build environments I agree could be easily moved to a different PR

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c (hunk 5 above)
Comment (markalle) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c:846]: okay, I'm making that change

**Comment by markalle** [comment]
I did a bunch of separations in to different PRs.  The others are

https://github.com/pmodels/mpich/pull/5099
https://github.com/pmodels/mpich/pull/5100
https://github.com/pmodels/mpich/pull/5101

File: src/mpl/src/mem/mpl_trmem.c (hunk 4 above)
Comment (markalle) [reply src/mpl/src/mem/mpl_trmem.c]: I've separated it now

File: src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c (hunk 3 above)
Comment (markalle) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c]: I separated it to different PRs

Comment (hzhou) [reply src/mpi/romio/adio/ad_gpfs/ad_gpfs_wrcoll.c]: Thanks. When we review or cherry-pick, we always do it on a per-PR basis. So it is desirable to make each PR with its obvious scope. It is not the number of PRs that inhibit reviewing. It is a finite amount of changes that we need to review one way or another. Having them separate allows us to review in a simpler context thus help making the process easier and faster.

**Comment by hzhou** [comment]
This PR has been replaced, ref: https://github.com/pmodels/mpich/pull/4948#issuecomment-790945664

**Comment by markalle** [comment]
Note: I opened a new https://github.com/pmodels/mpich/pull/5460 with the alltoallv fix, since it looks like that part got missed when this original PR was closed.  (This had been an oversized PR that I later separated out into multiple PRs).
